tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift

```diff
@@ -233,38 +237,277 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+                                    DWORD(FILE_MAP_ALL_ACCESS),
+                                    0,
+                                    0,
+                                    SIZE_T(bufSize));
+    if buf == LPVOID(bitPattern: 0) {
+      print("MapViewOfFile failed \(GetLastError())")
       return
     }
+    defer { UnmapViewOfFile(buf) }
 
-    repeat {
-      var entry: HEAPENTRY32 = HEAPENTRY32()
-      entry.dwSize = SIZE_T(MemoryLayout<HEAPENTRY32>.size)
+    // Set up the event objects
+    let (hReadEvent, hWriteEvent) = createEventPair(dwProcessId)
+    if hReadEvent == HANDLE(bitPattern: 0) {
+      print("CreateEvent failed \(GetLastError())")
+      return
+    }
+    defer { CloseHandle(hReadEvent) }
+    if hWriteEvent == HANDLE(bitPattern: 0) {
+      print("CreateEvent failed \(GetLastError())")
+      return
+    }
+    defer { CloseHandle(hWriteEvent) }
 
-      if !Heap32First(&entry, dwProcessId, heap.th32HeapID) {
-        // FIXME(compnerd) log error
-        continue
+    // Allocate the dll path string in the remote process.
+    guard let dllPathRemote = allocateDllPathRemote() else {
+      return
+    }
+
+    var hKernel32 = HMODULE(bitPattern: 0)
+    "kernel32".withCString() {
+      hKernel32 = GetModuleHandleA($0)
+    }
+
+    // Load the dll and start the heap walk
+    let hookAddr = unsafeBitCast(GetProcAddress(hKernel32, "LoadLibraryW"),
+                                 to: LPTHREAD_START_ROUTINE.self)
+    let hThread: HANDLE = CreateRemoteThread(self.process, nil, 0, hookAddr,
+                                             dllPathRemote, 0, nil)
+    if hThread == HANDLE(bitPattern: 0) {
+      print("CreateRemoteThread failed \(GetLastError())")
+      return
+    }
+    defer { CloseHandle(hThread) }
+
+    // The main heap iteration loop.
+    while true {
+      let wait = WaitForSingleObject(hReadEvent, waitTimeoutMs)
+      if wait != WAIT_OBJECT_0 {
+        print("WaitForSingleObject failed \(wait)")
+        return
       }
 
-      repeat {
-        if entry.dwFlags & DWORD(LF32_FREE) == DWORD(LF32_FREE) { continue }
-        body(swift_addr_t(entry.dwAddress), UInt64(entry.dwBlockSize))
-      } while Heap32Next(&entry)
+      let entryCount = bufSize/MemoryLayout<HeapEntry>.size
+      let buffer = buf.bindMemory(to: HeapEntry.self, capacity: 1)
+      let entry = buffer.pointee
+      if entry.Address == UInt.max {
+        // The buffer containing zero entries, indicated by the first entry
+        // contains -1, means, we are done. Break out of loop.
+        if !SetEvent(hWriteEvent) {
+          print("SetEvent failed \(GetLastError())")
+          return
+        }
+        break
+      }
+      for i in 0..<entryCount {
+        let entry = buffer.advanced(by: i).pointee
+        if entry.Address == 0 {
+          // Done. Break out of loop.
+          break
+        }
+        body(swift_addr_t(entry.Address), UInt64(entry.Size))
+      }
```

Comment:
I think that this could be a bit nicer as:

```swift
for entry in UnsafeMutableBufferPointer(start: buf.bindMemory(to: HeapEntry.self,
                                                              capacity: entryCount),
                                        count: entryCount) {
  if entry.Address == UInt.max {
    if !SetEvent(hWriteEvent) {
      print("SetEvent failed: \(GetLastError())")
      return
    }
    break
  }
  if entry.Address == 0 { break }
  body(swift_addr_t(entry.Address), UInt64(entry.Size))
}
```

This avoids the `advanced(by:)` dance.

Reply:
Done. Is there a way to check the iteration index in this for-in loop to have the UInt.max check is for the first entry only like `&entry == buf`?